src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,960 @@
+  const size_t col_size = this->options["col_size"];
+  // p_rad = radius of image patches.
+  // g_rad = radius of ground truth patches.
+  const size_t p_rad = p_size / 2, g_rad = g_size / 2;
+  
+  arma::vec FtrDim = this->GetFeatureDimension();
+  const size_t n_ftr_dim = FtrDim(0) + FtrDim(1);
+  const size_t n_smp_ftr_dim = (size_t)(n_ftr_dim * fraction);
+
+  for(size_t i = 0; i < n_tree; ++i)
+  {
+    //implement the logic for if data already exists.
+    MatType ftrs = arma::zeros(n_pos + n_neg, n_smp_ftr_dim);
+
+    //effectively a 3d array. . .
+    MatType lbls = arma::zeros( g_size * g_size, (n_pos + n_neg ));
+    // still to be done: store features and labels calculated 
+    // in the loop and store it in these Matrices.
+    // Could use some suggestions for this.
+
+    size_t loop_iter = num_images * 5;
+    for(size_t j = 0; j < loop_iter; j += 5)
+    {
+      MatType img, bnds, segs;
+      img = InputData.submat(j * row_size, 0, (j + 3) * row_size - 1, col_size - 1);
+      bnds = InputData.submat( (j + 3) * row_size, 0, \
+                        (j + 4) * row_size - 1, col_size - 1 );
+      segs = InputData.submat( (j + 4) * row_size, 0, \
+                        (j + 5) * row_size - 1, col_size - 1 );
+
+      MatType mask = arma::zeros(row_size, col_size);
+      for(size_t b = 0; b < mask.n_cols; b = b + shrink)
+        for(size_t a = 0; a < mask.n_rows; a = a + shrink)
+          mask(a, b) = 1;
+      mask.col(p_rad - 1).fill(0);
+      mask.row( (mask.n_rows - 1) - (p_rad - 1) ).fill(0);
+      mask.submat(0, 0, mask.n_rows - 1, p_rad - 1).fill(0);
+      mask.submat(0, mask.n_cols - p_rad, mask.n_rows - 1, 
+                  mask.n_cols - 1).fill(0);
+
+      // number of positive or negative patches per ground truth.
+      //int n_patches_per_gt = (int) (ceil( (float)n_pos / num_images ));
+      const size_t n_patches_per_gt = 500;
+      //cout << "n_patches_per_gt = " << n_patches_per_gt << endl;
+      MatType dis = arma::sqrt( this->DistanceTransformImage(bnds, 1) );
+      MatType dis2 = dis;
+      //dis.transform( [](double val, const int& g_rad) { return (double)(val < g_rad); } );
+      //dis2.transform( [](double val, const int& g_rad) { return (double)(val >= g_rad); } );
+      //dis.elem( arma::find(dis >= g_rad) ).zeros();
+      //dis2.elem( arma::find(dis < g_rad) ).zeros();
+      
+
+      arma::uvec pos_loc = arma::find( (dis < g_rad) % mask );
+      arma::uvec neg_loc = arma::find( (dis >= g_rad) % mask );
+
+      pos_loc = arma::shuffle(pos_loc);
+      neg_loc = arma::shuffle(neg_loc);
+
+      arma::umat loc(n_patches_per_gt * 2, 2);
+      //cout << "pos_loc size: " << arma::size(pos_loc) << " neg_loc size: " << arma::size(neg_loc) << endl;
+      //cout << "n_patches_per_gt = " << n_patches_per_gt << endl;
+      for(size_t i = 0; i < n_patches_per_gt; ++i)
+      {
+        loc.row(i) = arma::ind2sub(arma::size(dis.n_rows, dis.n_cols), pos_loc(i) ).t();
```

Comment:
Do you think, we could store the location in the columns, in this case we don't have to transpose the locations vector.